error.go

```diff
@@ -27,187 +27,99 @@ import (
-	multiline := w.Flag('+') && v == 'v'
-	verb := "%v"
-	if multiline {
-		verb = "%+v"
-	}
+var _ Error = errSpecification{}
 
-	// "context: " or "context:\n"
-	c.writeMessage(w, verb)
-	io.WriteString(w, ":")
-	if multiline {
-		io.WriteString(w, "\n")
-	} else {
-		io.WriteString(w, " ")
-	}
+func (e errSpecification) dummy() {}
 
-	fmt.Fprintf(w, verb, c.cause())
+func (e errSpecification) Unwrap() error {
+	return e.Cause
 }
 
-// RootCause returns the original error that caused the provided dig failure.
-//
-// RootCause may be used on errors returned by Invoke to get the original
-// error returned by a constructor or invoked function.
-func RootCause(err error) error {
-	for {
-		if e, ok := err.(causer); ok {
-			err = e.cause()
-		} else {
-			return err
-		}
-	}
+func (e errSpecification) writeMessage(w io.Writer, _ string) {
+	fmt.Fprintf(w, e.Message)
 }
 
-// errf is a version of fmt.Errorf with support for a chain of multiple
-// formatted error messages.
-//
-// After msg, N arguments are consumed as formatting arguments for that
-// message, where N is the number of % symbols in msg. Following that, another
-// string may be added to become the next error in the chain. Each new error
-// is the `cause()` for the prior error.
-//
-//	 err := errf(
-//	   "could not process %v", thing,
-//	   "name %q is invalid", thing.Name,
-//	)
-//	fmt.Println(err)  // could not process Thing: name Foo is invalid
-//	fmt.Println(RootCause(err))  // name Foo is invalid
-//
-// In place of a string, the last error can be another error, in which case it
-// will be treated as the cause of the prior error chain.
-//
-//	 errf(
-//	   "could not process %v", thing,
-//	   "date %q could not be parsed", thing.Date,
-//	   parseError,
-//	)
-func errf(msg string, args ...interface{}) error {
-	// By implementing buildErrf as a closure rather than a standalone
-	// function, we're able to ensure that it is called only from errf, or
-	// from itself (recursively). By controlling these invocations in such
-	// a tight space, we are able to easily verify manually that we
-	// checked len(args) > 0 before making the call.
-	var buildErrf func([]interface{}) error
-	buildErrf = func(args []interface{}) error {
-		arg, args := args[0], args[1:] // assume len(args) > 0
-		if arg == nil {
-			digerror.BugPanicf("arg must not be nil")
-		}
+func (e errSpecification) Error() string { return fmt.Sprint(e) }
+func (e errSpecification) Format(w fmt.State, c rune) {
+	formatError(e, w, c)
+}
 
-		switch v := arg.(type) {
-		case string:
-			need := numFmtArgs(v)
-			if len(args) < need {
-				digerror.BugPanicf("string %q needs %v arguments, got %v",
-					v,
-					need,
-					len(args))
-			}
+// wrappedError is a dig.Error with additional functionality
+// that allows them to be formatted via formatError()
+type wrappedError interface {
+	Error
+	fmt.Formatter
+	Unwrap() error
+	writeMessage(w io.Writer, v string)
+}
```

Comment:
I haven't gone through and verified every single dig.Error implementations to see if they implement wrappedError, but it seems like they do. In that case, can you just merge the definitions into one, and remove the dummy() method on dig.Error interface?